### ISSUE TITLE
An issue about the way to calculate the mean / error on one test env.

### ISSUE THREAD

**Issue by bad-meets-joke** [facebookresearch/DomainBed#139]
Hello, I have a question about the calculation method of the mean / error on one test env.

According to the code,  the calculation steps are

1. the test acc of one trial is from the hparams with max val acc across 20 different hparam sets in this trial.
2. the mean / error of test acc is calucated across multiple trials.

The steps expose an issue that **the test acc of different trials are from trained model with different hparam values**,  I think the test acc of each trial  should be based on the same hparams.

And the calculation steps **SHOULD** be
1. for each hparam set, calculate the mean of val acc across multiple trials, also the mean / error of test acc accross multiple trials
2. The final mean / error of test acc is from the hparam set with the max mean of val acc 

Welcome to point out my mistakes and looking forward to your reply. Thanks.

**Comment by piotr-teterwak** [comment]
Hi @bad-meets-joke , this is a design choice discussed in the paper (https://arxiv.org/pdf/2007.01434, Section 5, **Standard Error Bars**). The general idea is that the entire process of splitting the data, initializing the model, and doing the hyper-parameter search IS the learning algorithm, just that some parts are search-based (hyper-parameters) and some parts are optimization based (eg ERM). 

The reason this kind of reasoning is important is that it allows one to test the robustness of the whole system. For example, maybe one set of hyper-parameters is good for algorithm A, and another set of hyper-parameters is good for Algorithm B. Using your proposed methodology, we sample 20 sets of hyper-parameters. It's not unlikely that amongst those 20,  Algorithm A hits the jackpot hparams. Simply sampling the hyper-parameters once, and rerunning across multiple trials for each hparam set (as you propose) would unfairly show that Algorithm A is better.  

In summary, the error bars/mean show **how the system as whole** performs, not just the **optimization part**. 

Let me know if you have additional questions.

**Comment by piotr-teterwak** [comment]
Closing for now, feel free to reopen with more questions.
